modin/core/dataframe/algebra/default2pandas/groupby.py

```diff
@@ -270,6 +293,172 @@ def build_groupby(cls, func):
+            return False, [], []
+
+        if is_multi_by is None:
+            is_multi_by = len(internal_by_cols) > 1
+
+        # Pandas logic of resolving naming conflicts is the following:
+        #   1. If any categorical is in 'by' and 'by' is multi-column, then the categorical
+        #      index is prioritized: drop intersected columns and insert all of the 'by' index
+        #      levels to the frame as columns.
+        #   2. Otherwise, aggregation result is prioritized: drop intersected index levels and
+        #      insert the filtered ones to the frame as columns.
+        if by_cols_dtypes is not None:
+            keep_index_levels = (
+                is_multi_by
+                and selection is None
+                and any(isinstance(x, pandas.CategoricalDtype) for x in by_cols_dtypes)
+            )
+        else:
+            keep_index_levels = False
+
+        # 1. We insert 'by'-columns to the result at the beginning of the frame and so only to the
+        #    first partition, if partition_idx != 0 we just drop the index. If there are no columns
+        #    that are required to drop (keep_index_levels is True) then we can exit here.
+        # 2. We don't insert 'by'-columns to the result if 'by'-data came from a different
+        #    frame (drop is False), there's only one exception for this rule: if the `method` is "size",
+        #    so if (drop is False) and method is not "size" we just drop the index and so can exit here.
+        if (not keep_index_levels and partition_idx != 0) or (
+            not drop and method != "size"
+        ):
+            return True, [], []
+
+        if not isinstance(internal_by_cols, pandas.Index):
+            if not is_list_like(internal_by_cols):
+                internal_by_cols = [internal_by_cols]
+            internal_by_cols = pandas.Index(internal_by_cols)
+
+        internal_by_cols = (
+            internal_by_cols[~internal_by_cols.str.match(r"__reduced__.*", na=False)]
+            if hasattr(internal_by_cols, "str")
+            else internal_by_cols
+        )
+
+        if selection is not None:
+            selection = (
+                selection
+                if isinstance(selection, pandas.Index)
+                else pandas.Index(selection)
+            )
+
+        lvls_to_drop = []
+        cols_to_drop = []
+
+        if not keep_index_levels:
+            # We want to insert only these internal-by-cols that are not presented
+            # in the result in order to not create naming conflicts
+            cols_to_insert = (
+                internal_by_cols.copy()
+                if selection is None and internal_by_cols.nlevels != result_cols.nlevels
+                else frozenset(
+                    (
+                        col
+                        for col in internal_by_cols
+                        if col not in (result_cols if selection is None else selection)
+                    )
+                )
+            )
```

Comment:
I suggest to rewrite this in a normal `if..else` way, current one looks overcomplicated.

Don't try to fit everything in a single statement, it makes things harder to understand in a lot of cases.

```python
if selection is None:
    if internal_by_cols.nlevels != result_cols.nlevels:
        cols_to_insert = internal_by_cols.copy()
    else:
        cols_to_insert = frozenset(internal_by_cols) - frozenset(result_cols)
else:
    cols_to_insert = frozenset(internal_by_cols) - frozenset(selection)
```
or something along this lines